rfcs/001-mid-level-file-api.md

```diff
@@ -0,0 +1,194 @@
+# Mid-Level File Api
+
+## Summary
+
+This is a mid-level API wrapper around the various File related JavaScript apis. The possibility for a higher level API is left open.
+
+The mid-level file API aims to implement File I/O on top of the raw JS apis found here: https://w3c.github.io/FileAPI/ which includes listing files, creating files, and reading files.
+
+## The API
+
+First we have `FileList` which is simply a list of `File`s. This is an opaque struct that offers a way to iterate over `Files` and get individual files by index:
+
+```rust
+#[derive(Debug, Clone)]
+struct FileList { ... }
+
+impl FileList {
+  fn get(index: u32) -> Option<File> { ... }
+
+  fn len(&self) -> u32 { ... }
+
+  fn iter(&self) -> FileListIter { ... }
+
+  fn as_raw(&self) -> &web_sys::FileList { ... }
+}
+
+impl AsRef<web_sys::FileList> for FileList { ... }
+impl From<web_sys::FileList> for FileList { ... }
+impl Index<usize> for FileList {
+    type Output = File;
+    fn index(&self, index: usize) -> &Self::Output { ... }
+}
+```
+
+There is no `FileList::new` since creating a raw `web_sys::FileList` is not possible without going through a `web_sys::HtmlInputElement`.
+
+Next we the trait `BlobLike`
+```rust
+trait BlobLike {
+    fn size(&self) -> u64 { ... }
+
+    // the mime crate and mimes from blobs both conform to rfc6838
+    #[cfg(feature = "mime")]
+    fn mime_type(&self) -> Result<mime::Mime, mime::FromStrError> { ... }
+
+    fn raw_mime_type(&self) -> String { ... }
+
+    fn as_raw(&self) -> &web_sys::Blob;
+}
+```
+There are two structs that implement this trait: `Blob` and `File`.
+
+```rust
+#[derive(Debug, Clone)]
+struct Blob { ... }
+
+impl Blob {
+  fn new<T>(content: Option<T>) -> Blob
+    where
+        T: std::convert::Into<BlobContents> // We'll look at BlobContents below
+    { ... }
+
+  fn new_with_options<T>(content: Option<T>, mime_type: String) -> Blob
+    where
+        T: std::convert::Into<BlobContents>
+    { ... }
+
+  fn slice(&self, start: u64, end: u64) -> Blob { ... }
+}
+
+impl From<web_sys::Blob> for Blob { ... }
+
+impl BlobLike for Blob { ... }
+
+#[derive(Debug, Clone)]
+pub struct File { ... }
+
+impl File {
+  fn new<T>(
+        name: String,
+        contents: Option<T>,
+  ) -> File
+    where
+        T: std::convert::Into<BlobContents>,
+    { ... }
+
+  fn new_with_options<T>(
+        name: String,
+        contents: Option<T>,
+        mime_type: Option<String>,
+        last_modified_date: Option<u64>,
+  ) -> File
+    where
+        T: std::convert::Into<BlobContents>,
+    { ... }
+
+  fn name(&self) -> String { ... }
+
+  fn last_modified_since_epoch(&self) -> Duration { ... }
+
+  fn slice(&self, start: u64, end: u64) -> File { ... }
+
+  fn as_blob(&self) -> Blob { ... }
```

Comment:
Is this needed, since `BlobLike` defines `as_raw`?

Reply:
Ah, I see the difference. This returns a `Blob`, whereas `as_raw` returns a `web_sys::Blob`.

I'm still not a big fan of this method though. It feels too much like it should be something else, like `Deref<Target = Blob>`, or `AsRef`, or `Into`, or...

Reply:
`AsRef` and `Deref` don't work because this isn't returning a reference.

Since the conversion's ownership is from borrowed to owned, it should probably be called `to_blob` to follow [the C-CONV guidelines](https://rust-lang-nursery.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv).

Reply:
@rylev @Pauan how does renaming to `to_blob` sound?

Reply:
That sounds fine to me.

Reply:
Yeah, I like `to_blob` better.